File: app/src/main/java/info/nightscout/androidaps/plugins/aps/loop/LoopFragment.java
package info.nightscout.androidaps.plugins.aps.loop;


import android.app.Activity;
import android.os.Bundle;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import com.squareup.otto.Subscribe;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import info.nightscout.androidaps.MainApp;
import info.nightscout.androidaps.R;
import info.nightscout.androidaps.interfaces.Constraint;
import info.nightscout.androidaps.plugins.aps.loop.events.EventLoopSetLastRunGui;
import info.nightscout.androidaps.plugins.aps.loop.events.EventLoopUpdateGui;
import info.nightscout.androidaps.plugins.common.SubscriberFragment;
import info.nightscout.androidaps.utils.FabricPrivacy;

public class LoopFragment extends SubscriberFragment {
    @BindView(R.id.loop_run)
    Button runNowButton;
    @BindView(R.id.loop_lastrun)
    TextView lastRunView;
    @BindView(R.id.loop_lastenact)
    TextView lastEnactView;
    @BindView(R.id.loop_source)
    TextView sourceView;
    @BindView(R.id.loop_request)
    TextView requestView;
    @BindView(R.id.loop_constraintsprocessed)
    TextView constraintsProcessedView;
    @BindView(R.id.loop_constraints)
    TextView constraintsView;
    @BindView(R.id.loop_tbrsetbypump)
    TextView tbrSetByPumpView;
    @BindView(R.id.loop_smbsetbypump)
    TextView smbSetByPumpView;


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        try {
            View view = inflater.inflate(R.layout.loop_fragment, container, false);
            unbinder = ButterKnife.bind(this, view);
            return view;
        } catch (Exception e) {
            FabricPrivacy.logException(e);
        }
        return null;
    }

    @OnClick(R.id.loop_run)
    void onRunClick() {
        lastRunView.setText(MainApp.gs(R.string.executing));
        new Thread(() -> LoopPlugin.getPlugin().invoke("Loop button", true)).start();
    }

    @Subscribe
    public void onStatusEvent(final EventLoopUpdateGui ev) {
        updateGUI();
    }

    @Subscribe
    public void onStatusEvent(final EventLoopSetLastRunGui ev) {
        clearGUI();
        final Activity activity = getActivity();
        if (activity != null)
            activity.runOnUiThread(() -> {
                synchronized (LoopFragment.this) {
                    if (lastRunView != null) lastRunView.setText(ev.text);
                }
            });
    }


    @Override
    protected void updateGUI() {
        Activity activity = getActivity();
        if (activity != null)
            activity.runOnUiThread(() -> {
                synchronized (LoopFragment.this) {
                    if (!isBound()) return;
                    LoopPlugin.LastRun lastRun = LoopPlugin.lastRun;
                    if (lastRun != null) {
                        requestView.setText(lastRun.request != null ? lastRun.request.toSpanned() : "");
                        constraintsProcessedView.setText(lastRun.constraintsProcessed != null ? lastRun.constraintsProcessed.toSpanned() : "");
                        sourceView.setText(lastRun.source != null ? lastRun.source : "");
                        lastRunView.setText(lastRun.lastAPSRun != null && lastRun.lastAPSRun.getTime() != 0 ? lastRun.lastAPSRun.toLocaleString() : "");
                        lastEnactView.setText(lastRun.lastEnact != null && lastRun.lastEnact.getTime() != 0 ? lastRun.lastEnact.toLocaleString() : "");
                        tbrSetByPumpView.setText(lastRun.tbrSetByPump != null ? Html.fromHtml(lastRun.tbrSetByPump.toHtml()) : "");
                        smbSetByPumpView.setText(lastRun.smbSetByPump != null ? Html.fromHtml(lastRun.smbSetByPump.toHtml()) : "");

                        String constraints = "";
                        if (lastRun.constraintsProcessed != null) {
                            Constraint<Double> allConstraints = new Constraint<>(0d);
                            if (lastRun.constraintsProcessed.rateConstraint != null)
                                allConstraints.copyReasons(lastRun.constraintsProcessed.rateConstraint);
                            if (lastRun.constraintsProcessed.smbConstraint != null)
                                allConstraints.copyReasons(lastRun.constraintsProcessed.smbConstraint);
                            constraints = allConstraints.getMostLimitedReasons();
                        }
                        constraintsView.setText(constraints);
                    }
                }
            });
    }

    void clearGUI() {
        Activity activity = getActivity();
        if (activity != null)
            activity.runOnUiThread(() -> {
                synchronized (LoopFragment.this) {
                    if (isBound()) {
                        requestView.setText("");
                        constraintsProcessedView.setText("");
                        sourceView.setText("");
                        lastRunView.setText("");
                        lastEnactView.setText("");
                        tbrSetByPumpView.setText("");
                        smbSetByPumpView.setText("");
                    }
                }
            });
    }

    boolean isBound() {
        return requestView != null
                && constraintsProcessedView != null
                && sourceView != null
                && lastRunView != null
                && lastEnactView != null
                && tbrSetByPumpView != null
                && smbSetByPumpView != null
                && constraintsView != null
                && runNowButton != null;
    }
}

File: app/src/main/java/info/nightscout/androidaps/plugins/aps/loop/LoopFragment.kt
package info.nightscout.androidaps.plugins.aps.loop

import android.os.Bundle
import android.view.*
import dagger.android.support.DaggerFragment
import info.nightscout.androidaps.R
import info.nightscout.androidaps.databinding.LoopFragmentBinding
import info.nightscout.androidaps.interfaces.Constraint
import info.nightscout.androidaps.interfaces.Loop
import info.nightscout.androidaps.plugins.aps.loop.events.EventLoopSetLastRunGui
import info.nightscout.androidaps.plugins.aps.loop.events.EventLoopUpdateGui
import info.nightscout.androidaps.plugins.bus.RxBus
import info.nightscout.androidaps.utils.DateUtil
import info.nightscout.androidaps.utils.FabricPrivacy
import info.nightscout.androidaps.utils.HtmlHelper
import info.nightscout.androidaps.interfaces.ResourceHelper
import info.nightscout.androidaps.utils.rx.AapsSchedulers
import info.nightscout.shared.logging.AAPSLogger
import info.nightscout.shared.sharedPreferences.SP
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.kotlin.plusAssign
import javax.inject.Inject

class LoopFragment : DaggerFragment() {

    @Inject lateinit var aapsLogger: AAPSLogger
    @Inject lateinit var aapsSchedulers: AapsSchedulers
    @Inject lateinit var rxBus: RxBus
    @Inject lateinit var sp: SP
    @Inject lateinit var rh: ResourceHelper
    @Inject lateinit var fabricPrivacy: FabricPrivacy
    @Inject lateinit var loop: Loop
    @Inject lateinit var dateUtil: DateUtil

    private val ID_MENU_RUN = 1

    private var disposable: CompositeDisposable = CompositeDisposable()

    private var _binding: LoopFragmentBinding? = null

    // This property is only valid between onCreateView and
    // onDestroyView.
    private val binding get() = _binding!!

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        setHasOptionsMenu(true)
        _binding = LoopFragmentBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        with(binding.swipeRefresh) {
            setColorSchemeColors(rh.gac(context, R.attr.colorPrimaryDark), rh.gac(context, R.attr.colorPrimary), rh.gac(context, R.attr.colorSecondary))
            setOnRefreshListener {
                binding.lastrun.text = rh.gs(info.nightscout.androidaps.R.string.executing)
                Thread { loop.invoke("Loop swiperefresh", true) }.start()
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        if (isResumed) {
            menu.removeItem(ID_MENU_RUN)
            menu.add(Menu.FIRST, ID_MENU_RUN, 0, rh.gs(R.string.openapsma_run)).setShowAsAction(MenuItem.SHOW_AS_ACTION_NEVER)
            menu.setGroupDividerEnabled(true)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean =
        when (item.itemId) {
            ID_MENU_RUN -> {
                binding.lastrun.text = rh.gs(R.string.executing)
                Thread { loop.invoke("Loop menu", true) }.start()
                true
            }

            else        -> false
        }

    @Synchronized
    override fun onResume() {
        super.onResume()
        disposable += rxBus
            .toObservable(EventLoopUpdateGui::class.java)
            .observeOn(aapsSchedulers.main)
            .subscribe({
                           updateGUI()
                       }, fabricPrivacy::logException)

        disposable += rxBus
            .toObservable(EventLoopSetLastRunGui::class.java)
            .observeOn(aapsSchedulers.main)
            .subscribe({
                           clearGUI()
                           binding.lastrun.text = it.text
                       }, fabricPrivacy::logException)

        updateGUI()
        sp.putBoolean(R.string.key_objectiveuseloop, true)
    }

    @Synchronized
    override fun onPause() {
        super.onPause()
        disposable.clear()
    }

    @Synchronized
    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    @Synchronized
    fun updateGUI() {
        if (_binding == null) return
        loop.lastRun?.let {
            binding.request.text = it.request?.toSpanned() ?: ""
            binding.constraintsprocessed.text = it.constraintsProcessed?.toSpanned() ?: ""
            binding.source.text = it.source ?: ""
            binding.lastrun.text = dateUtil.dateAndTimeString(it.lastAPSRun)
            binding.smbrequestTime.text = dateUtil.dateAndTimeAndSecondsString(it.lastSMBRequest)
            binding.smbexecutionTime.text = dateUtil.dateAndTimeAndSecondsString(it.lastSMBEnact)
            binding.tbrrequestTime.text = dateUtil.dateAndTimeAndSecondsString(it.lastTBRRequest)
            binding.tbrexecutionTime.text = dateUtil.dateAndTimeAndSecondsString(it.lastTBREnact)

            binding.tbrsetbypump.text = it.tbrSetByPump?.let { tbrSetByPump -> HtmlHelper.fromHtml(tbrSetByPump.toHtml()) }
                ?: ""
            binding.smbsetbypump.text = it.smbSetByPump?.let { smbSetByPump -> HtmlHelper.fromHtml(smbSetByPump.toHtml()) }
                ?: ""

            val constraints =
                it.constraintsProcessed?.let { constraintsProcessed ->
                    val allConstraints = Constraint(0.0)
                    constraintsProcessed.rateConstraint?.let { rateConstraint -> allConstraints.copyReasons(rateConstraint) }
                    constraintsProcessed.smbConstraint?.let { smbConstraint -> allConstraints.copyReasons(smbConstraint) }
                    allConstraints.getMostLimitedReasons(aapsLogger)
                } ?: ""
            binding.constraints.text = constraints
            binding.swipeRefresh.isRefreshing = false
        }
    }

    @Synchronized
    private fun clearGUI() {
        binding.request.text = ""
        binding.constraints.text = ""
        binding.constraintsprocessed.text = ""
        binding.source.text = ""
        binding.lastrun.text = ""
        binding.smbrequestTime.text = ""
        binding.smbexecutionTime.text = ""
        binding.tbrrequestTime.text = ""
        binding.tbrexecutionTime.text = ""
        binding.tbrsetbypump.text = ""
        binding.smbsetbypump.text = ""
        binding.swipeRefresh.isRefreshing = false
    }
}
